File: backend/__SLUG___node/src/main.blank.ts
import express, { type NextFunction, type Request, type Response } from "express";
import { Pool } from "pg";

import { initLogger, getRootLogger } from "./logger/logger";
import { LoggingDB } from "./logger/db";
import { trace } from "./middleware/trace";
import { requestLog } from "./middleware/requestLog";

import * as health from "./domains/health/routes";

const API_BASE_PATH = "/api/__SLUG__";

function initDbPool(): Pool {
  const connectionString = process.env.DATABASE_URL;
  if (connectionString) {
    return new Pool({ connectionString });
  }
  return new Pool({
    host: process.env.DB_HOST || "localhost",
    port: Number(process.env.DB_PORT || 5432),
    user: process.env.DB_USER || "__SLUG__",
    password: process.env.DB_PASS || "__SLUG___local_pass",
    database: process.env.DB_NAME || "__SLUG___local",
  });
}

function corsMiddleware(req: Request, res: Response, next: NextFunction): void {
  res.header("Access-Control-Allow-Origin", "*");
  res.header("Access-Control-Allow-Credentials", "true");
  res.header("Access-Control-Allow-Headers", "*");
  res.header("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS, PATCH");
  res.header("Access-Control-Expose-Headers", "X-Span-Id");
  if (req.method === "OPTIONS") {
    res.sendStatus(200);
    return;
  }
  next();
}

async function main(): Promise<void> {
  initLogger();
  const log = getRootLogger();

  const pool = initDbPool();
  void new LoggingDB(pool);

  const app = express();
  app.use(trace);
  app.use(requestLog);
  app.use(corsMiddleware);
  app.use(express.json());

  const apiRouter = express.Router();
  app.use(API_BASE_PATH, apiRouter);
  apiRouter.use(health.registerRoutes());

  app.use((err: Error, _req: Request, res: Response, _next: NextFunction) => {
    log.error({ err }, "unhandled error");
    res.status(500).json({ error: "internal server error" });
  });

  const port = Number(process.env.PORT || 8080);
  app.listen(port, () => {
    log.info({ port }, "starting server");
  });
}

main().catch((err) => {
  getRootLogger().fatal({ err }, "failed to start server");
  process.exit(1);
});

File: backend/__SLUG__/src/main.blank.go
package main

import (
	"database/sql"
	"go_backend/src/domains/health"
	"go_backend/src/logger"
	appMiddleware "go_backend/src/middleware"
	"net/http"
	"os"

	"github.com/gorilla/mux"
	_ "github.com/lib/pq"
	"go.uber.org/zap"
)

func main() {
	logger.InitLogger()

	rawDB, err := initDB()
	if err != nil {
		zap.L().Fatal("failed to connect to database", zap.Error(err))
	}
	defer rawDB.Close()

	_ = logger.NewLoggingDB(rawDB)

	router := mux.NewRouter()
	router.Use(appMiddleware.Trace)
	router.Use(appMiddleware.RequestLog)
	router.Use(corsMiddleware)

	apiRouter := router.PathPrefix("/api/__SLUG__").Subrouter()

	healthRouter := health.RegisterRoutes()
	apiRouter.PathPrefix("/health").Handler(http.StripPrefix("/api/__SLUG__", healthRouter))

	port := os.Getenv("PORT")
	if port == "" {
		port = "8080"
	}

	zap.L().Info("starting server", zap.String("port", port))
	if err := http.ListenAndServe(":"+port, router); err != nil {
		zap.L().Fatal("server failed", zap.Error(err))
	}
}

func initDB() (*sql.DB, error) {
	dsn := os.Getenv("DATABASE_URL")
	if dsn == "" {
		host := os.Getenv("DB_HOST")
		if host == "" {
			host = "localhost"
		}
		port := os.Getenv("DB_PORT")
		if port == "" {
			port = "5432"
		}
		user := os.Getenv("DB_USER")
		if user == "" {
			user = "__SLUG__"
		}
		pass := os.Getenv("DB_PASS")
		if pass == "" {
			pass = "__SLUG___local_pass"
		}
		name := os.Getenv("DB_NAME")
		if name == "" {
			name = "__SLUG___local"
		}
		dsn = "host=" + host + " port=" + port + " user=" + user + " password=" + pass + " dbname=" + name + " sslmode=disable"
	}
	db, err := sql.Open("postgres", dsn)
	if err != nil {
		return nil, err
	}
	if err := db.Ping(); err != nil {
		return nil, err
	}
	return db, nil
}

func corsMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Access-Control-Allow-Origin", "*")
		w.Header().Set("Access-Control-Allow-Credentials", "true")
		w.Header().Set("Access-Control-Allow-Headers", "*")
		w.Header().Set("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS, PATCH")
		w.Header().Set("Access-Control-Expose-Headers", "X-Span-Id")
		if r.Method == "OPTIONS" {
			w.WriteHeader(http.StatusOK)
			return
		}
		next.ServeHTTP(w, r)
	})
}
